crates/biome_js_analyze/src/services/semantic_class.rs

```diff
@@ -1,25 +1,149 @@
-    JsPreUpdateExpression, JsPropertyClassMember, JsStaticMemberAssignment,
-    JsStaticMemberExpression, JsSyntaxKind, JsSyntaxNode, JsVariableDeclarator, TextRange,
-    TsPropertyParameter,
+    AnyJsClassMember, AnyJsExpression, AnyJsRoot, JsArrayAssignmentPattern,
+    JsArrowFunctionExpression, JsAssignmentExpression, JsClassDeclaration, JsClassMemberList,
+    JsConstructorClassMember, JsFunctionBody, JsLanguage, JsObjectAssignmentPattern,
+    JsObjectBindingPattern, JsPostUpdateExpression, JsPreUpdateExpression, JsPropertyClassMember,
+    JsStaticMemberAssignment, JsStaticMemberExpression, JsSyntaxKind, JsSyntaxNode,
+    JsVariableDeclarator, TextRange, TsPropertyParameter,
 };
 
-use biome_analyze::QueryMatch;
+use biome_analyze::{
+    AddVisitor, FromServices, Phase, Phases, QueryKey, QueryMatch, Queryable, RuleKey,
+    RuleMetadata, ServiceBag, ServicesDiagnostic, Visitor, VisitorContext, VisitorFinishContext,
+};
 use biome_rowan::{
     AstNode, AstNodeList, AstSeparatedList, SyntaxNode, Text, WalkEvent, declare_node_union,
 };
 use std::collections::HashSet;
 
+#[derive(Clone)]
+pub struct SemanticClassServices {
+    class_member_references: ClassMemberReferences,
+}
+
+impl SemanticClassServices {
+    pub fn class_member_references(&self) -> &ClassMemberReferences {
+        &self.class_member_references
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ClassMemberReferencesService {
+    pub references: ClassMemberReferences,
+}
+
+impl FromServices for SemanticClassServices {
+    fn from_services(
+        rule_key: &RuleKey,
+        _rule_metadata: &RuleMetadata,
+        services: &ServiceBag,
+    ) -> biome_diagnostics::Result<Self, ServicesDiagnostic> {
+        let service: &ClassMemberReferencesService = services.get_service().ok_or_else(|| {
+            ServicesDiagnostic::new(rule_key.rule_name(), &["ClassMemberReferencesService"])
+        })?;
+        Ok(Self {
+            class_member_references: service.references.clone(),
+        })
+    }
+}
+
+impl Phase for SemanticClassServices {
+    fn phase() -> Phases {
+        Phases::Semantic
+    }
+}
+
+pub struct ClassMemberReferencesVisitor {
+    references: ClassMemberReferences,
+}
+
+impl ClassMemberReferencesVisitor {
+    pub(crate) fn new(_root: &AnyJsRoot) -> Self {
+        Self {
+            references: ClassMemberReferences::default(),
+        }
+    }
+}
+
+impl Visitor for ClassMemberReferencesVisitor {
+    type Language = JsLanguage;
+
+    fn visit(&mut self, event: &WalkEvent<JsSyntaxNode>, _ctx: VisitorContext<JsLanguage>) {
+        if let WalkEvent::Enter(node) = event
+            && let Some(js_class_declaration) = JsClassDeclaration::cast_ref(node)
+        {
+            let class_member_list = js_class_declaration.members();
+            let refs = class_member_references(&class_member_list);
+            self.references.reads.extend(refs.reads);
+            self.references.writes.extend(refs.writes);
+        }
```

Comment:
This is incorrect now. When we visit a CST, there might be multiple `class` in the same root. The service should track all `class` found in the root. The way it is implemented now, all references get mixed up together, risking causing confusion.

At the moment, the rule seems to work, but I'm not sure why and how. However, as you can notice, ALL reads and writes are mixed together. Is that right?